### PR TITLE
Fix meta and cleanup

### PR DESCRIPTION
**Pull request by francoisbessette**
List of changes:
-fixed bug where a new user would get blank meta-data (ex: phone number)
 when his record was created. I believe the next run of the plugin was 
 healing the user meta-data.

-fixed bug where local DB was not updating the field that a user deleted
 on the national web site. Ex: if user decided to no longer list his cell
 phone.

-only update users if needed. This solves problem where the Simple History
 plugin would log hundreds of NOP updates.

-give a more meaningful summary at the end of processing.

-clearer logging messages

-add log_dual to avoid duplication of logging statements all over code.

-fix bug where all runs of the plugin falling within the same hour
 would be logged into the same file, and a run could be split between
 2 files if it happened to cross a 1-hour boundary. Now every run of the 
 plugin is a different file, as long as there is one minute between them.
 This allows for easier development where you run the plugin, check the
 result, do a bit more code, run again.

-improve efficiency of logging function. Use a static variable to avoid
 searching the whole directory each time the log function is called. 

-Fix PHP warnings in the case where some received fields like Cell Phone are empty 

-Fix case where a user has 2 memberships. Pick the one with the latest expiry, and never update with an earlier expiry

### CODE REVIEW

File: admin/class-acc_user_importer-admin.php [hunk 1]
```diff
@@ -1,6 +1,7 @@
 <?php
 
 
+$logstr = "";		//handy global to store log string
```
Comment (cloetzi) [inline admin/class-acc_user_importer-admin.php]: Why does this need to be global? I think as it stands now, you could also use a class property instead.

In any case, we should probably prefix everything global to avoid accidentally interfering with WP or a plug-in. E.g. `$acc_logstr`.

File: admin/class-acc_user_importer-admin.php [hunk 2]
```diff
@@ -358,148 +374,192 @@ private function proccess_user_data ( $users ) {
...
 				
-				
 				//if email exists in different contact ID, skip updating user
-				if ( is_numeric($user_id) ) {	
-					$api_response['log'] .= "&nbsp;&gt; duplicate email found on another user <em>(user #" . $user_id . ")</em><br/>";
-					$api_response['log'] .= "&nbsp;&gt; user update skipped<br/>";
-					$this->log_local_output(" > duplicate email found on another user (user #" . $user_id . ")");
-					$this->log_local_output(" > user update skipped");
+				if ( is_numeric($user_id) ) {
+					$user2 = get_user_by( 'id', $user_id);
```
Comment (cloetzi) [inline admin/class-acc_user_importer-admin.php]: Just FYI, you could also use `get_userdata($user_id)`.

File: admin/class-acc_user_importer-admin.php [hunk 3]
```diff
@@ -358,148 +374,192 @@ private function proccess_user_data ( $users ) {
 			
 			//check if ID already exist
 			$user_id = username_exists($user_contact_id);
-			
+			$update_this_user = false;
+			$user_already_exists = false;
+			$create_this_user = false;
```
Comment (cloetzi) [inline admin/class-acc_user_importer-admin.php]: With all the flags, the code is pretty hard to follow. It works fine for now, but I'll hope to refactor that a bit at some point to not rely on the state of three flags. In my experience, it is pretty easy to introduce bugs in the future by losing track of one of the flags when doing changes.

File: admin/class-acc_user_importer-admin.php [hunk 4]
```diff
@@ -358,148 +374,192 @@ private function proccess_user_data ( $users ) {
...
+				if (!$update_this_user) {
+					if ($user_email != $user_meta->user_email) {
+						$this->log_dual(" > email changed from " . $user_meta->user_email . " to " . $user_email);
+						$update_this_user = true;
+					}
+				}
+
+				//Check if HomePhone changed
+				if (!$update_this_user) {
+					if ($user["HomePhone"] != $user_meta->home_phone) {
```
Comment (cloetzi) [inline admin/class-acc_user_importer-admin.php]: Without a check, if the "HomePhone" key exists, this will throw PHP notices whenever it is not existing. Not the end of the world, but it will flood logs and error tracking tools. We also need to consider the case if the metadata does not exist yet, as far as I can tell, then `$user_meta->home_phone` returns an empty string.

To avoid accessing an empty index, you could leverage the [Null coalescing operator](https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op) and write something like this:
```php
if (($homePhone = $user["HomePhone"] ?? '') != $user_meta->home_phone) {
  $this->log_dual(" > home phone changed from " . $user_meta->home_phone . " to " . $homePhone);
  $update_this_user = true;
}
```
That said, ideally, we'd re-use the `$homePhone` variable for updating the value as well, but that would be a bit convoluted as it stands now. I like the logic generally to only update if required, but I think the implementation can be simplified quite a bit. I am fine merging this first, then I give it a shot on top of your implementation, or directly committing to your branch. Your choice :slightly_smiling_face:

File: admin/class-acc_user_importer-admin.php [hunk 5]
```diff
@@ -358,148 +374,192 @@ private function proccess_user_data ( $users ) {
...
+			if ($update_this_user) {
+				//update core info (name, email)
+				$wp_user = wp_insert_user( $user_data );
+				if ( is_wp_error($wp_user) ) {
+					$this->log_dual(" > failed to update user");
+					$this->log_dual(" > WP:" . $wp_user->get_error_message());
+
+				} else {
+					//update user meta
+					update_user_meta( $wp_user, 'home_phone', $user['HomePhone'] );
```
Comment (cloetzi) [inline admin/class-acc_user_importer-admin.php]: Again, this will produce notices if the `HomePhone` (or any other index) is missing.

File: admin/class-acc_user_importer-admin.php [hunk 6]
```diff
@@ -635,7 +695,9 @@ public function enqueue_scripts() {
 	}
 	
 	public function log_local_output( $v ) {
```
Comment (cloetzi) [inline admin/class-acc_user_importer-admin.php:691]: Generally speaking, do we want to have a separate log file for each run? If so, I am wondering if we could leverage the access token we receive from the ACC API plus the start time of the run to have a 'known' log file name. That way we would not need any logic to check if a file exists or not, we just know the file name for the run and that's it.

Of course, this would require that we'd have to pass that information down to the backend for frontend initiated runs. That said, I'd not recommend doing this refactoring in this PR but in a future change.

File: admin/class-acc_user_importer-admin.php (hunk 1 above)
Comment (francoisbessette) [reply admin/class-acc_user_importer-admin.php]: Great comment, thanks!   Yeah, I had a feeling it was not super. So tell me, is the current scope just for the file, or it is global across all wordpress installation?

My goal was to enhance the code readability. There used to be 2 different statements for every log. Now there is only one, called log_dual.  I wanted to keep the invocation simple and did not want to pass the string back and forth. This is why I used the global.  I originally wanted to put it underneath the 'class acc_user_importer_Admin', but somehow it did not work. Forgive my lack of expertise with PHP and object oriented programming.

For now I will rename the variable as you suggested, and if you know how to use a class property, please go ahead and do that (since I am not sure what you are referring to).

File: admin/class-acc_user_importer-admin.php (hunk 2 above)
Comment (francoisbessette) [reply admin/class-acc_user_importer-admin.php]: Thanks, fixed!

File: admin/class-acc_user_importer-admin.php (hunk 4 above)
Comment (francoisbessette) [reply admin/class-acc_user_importer-admin.php]: OK, I will let you do the changes.

File: admin/class-acc_user_importer-admin.php (hunk 6 above)
Comment (francoisbessette) [reply admin/class-acc_user_importer-admin.php:691]: I agree with your proposal, it would be simpler and better. Later.

File: admin/class-acc_user_importer-admin.php (hunk 5 above)
Comment (francoisbessette) [reply admin/class-acc_user_importer-admin.php]: I will let you do the changes.

File: admin/class-acc_user_importer-admin.php [hunk 7]
```diff
@@ -358,148 +374,192 @@ private function proccess_user_data ( $users ) {
...
+
+				// Get all the user roles as an array.
+				$user_roles = $user_meta->roles;
+				// Check if user has expired or ex-member role. If so, update role to default
+				if ( in_array( $role_expiry1, $user_roles, true ) ) {
+					$this->log_dual(" > User is " . $role_expiry1 . ", refreshing role to " . $default_role);
+					$update_this_user = true;
+					$user_data["role"] = $default_role;
+					$role_refreshed[] = $user_contact_id;
+					$role_refreshed_email[] = $user_email;
```
Comment (francoisbessette) [inline admin/class-acc_user_importer-admin.php]: I have 2 arrays (roles_refreshed, role_refreshed_email) for storing the names and emails of those we refreshed roles.  I have a feeling there is a better way to do it. Feel free to improve.

File: admin/class-acc_user_importer-admin.php [hunk 8]
```diff
@@ -358,148 +374,192 @@ private function proccess_user_data ( $users ) {
...
+				//Check if Cell phone changed
+				if (!$update_this_user) {
+					if ($user["Cell Phone"] != $user_meta->cell_phone) {
+						$this->log_dual(" > cell phone changed from " . $user_meta->cell_phone . " to " . $user["Cell Phone"]);
+						$update_this_user = true;
+					}
+				}
+
+				//Check if MEMBERSHIP_N changed
+				if (!$update_this_user) {
```
Comment (francoisbessette) [inline admin/class-acc_user_importer-admin.php]: I decided to remove the above if.  This way the logic is simpler and the log will indicate all fields that changed for a given user, not only the first one.

File: admin/class-acc_user_importer-admin.php (hunk 3 above)
Comment (francoisbessette) [reply admin/class-acc_user_importer-admin.php]: Good comment. I restructured the code, I now only have 1 flag.  Much cleaner. I hope you like it.

File: admin/class-acc_user_importer-admin.php (hunk 4 above)
Comment (francoisbessette) [reply admin/class-acc_user_importer-admin.php]: I did the changes

File: admin/class-acc_user_importer-admin.php (hunk 1 above)
Comment (cloetzi) [reply admin/class-acc_user_importer-admin.php]: >  So tell me, is the current scope just for the file, or it is global across all wordpress installation?

That is a good question but I am not 100% sure. My understanding is that it is for the full Wordpress installation. This is true for sure for everything within `$GLOBAL` since that is a [suberglobal](https://www.php.net/manual/en/language.variables.superglobals.php). 
For the var itself, it is a [single scope](https://www.php.net/manual/en/language.variables.scope.php) which is limited, but Wordpress does some interesting things in how it loads files and plugins, so I am not sure how widely it is accessible.

File: admin/class-acc_user_importer-admin.php [hunk 9]
```diff
@@ -315,191 +326,234 @@ private function proccess_user_data ( $users ) {
...
+				$this->log_dual(" > user not found");
+
+				$user_id = email_exists($userEmail);
+				if ( is_numeric($user_id) ) {
+					//An existing user already has this email address, skip updating
+					$user2 = get_userdata($user_id);
+					$username2 = $user2->user_firstname . " " . $user2->user_lastname;
+					$this->log_dual(" > existing user #" . $user_id . " " . $username2 .
+					                " already has email " . $userEmail);
+					$this->log_dual(" > user update skipped");
```
Comment (cloetzi) [inline admin/class-acc_user_importer-admin.php:484]: My understanding is that this is not always correct. If `update_this_user` is set, we still might update it, right?

File: admin/class-acc_user_importer-admin.php (hunk 3 above)
Comment (cloetzi) [reply admin/class-acc_user_importer-admin.php]: Yes, much cleaner, thank you!

**Review by cloetzi — APPROVED** [review]
It looks good to me, but disclaimer, I did not test it on our side yet. I think it is good to merge and if there are bugs we can fix them in a subsequent PR. Thanks for all the improvements!

File: admin/class-acc_user_importer-admin.php (hunk 9 above)
Comment (francoisbessette) [reply admin/class-acc_user_importer-admin.php:484]: The only way to get here is with update_this_user being false. So we will not update it.
